conjure-go-contract/codecs/gzip.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2018 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codecs
+
+import (
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+)
+
+var _ Codec = codecGZIP{}
+
+// GZIP wraps an existing Codec and uses gzip for compression and decompression.
+func GZIP(codec Codec) Codec {
```

Comment:
The other codecs work standalone but this one requires a delegate codec. Could you add a comment on why that is that and how does gzip interact with the other plugin? Would it be better to add a sane default delegate codec and a new function that allows overriding it (e.g. `GZIPWithDelegate`)?

Reply:
So this new GZIP compression codec copies the form of the other compression codecs we already have in CGR:
- ZLib: https://github.com/palantir/conjure-go-runtime/blob/develop/conjure-go-contract/codecs/zlib.go
- Snappy: https://github.com/palantir/conjure-go-runtime/blob/develop/conjure-go-contract/codecs/snappy.go

The way these codecs work is by wrapping an encoder (i.e plaintext, protobuf) and taking whatever those codecs return, running compression, and returning the result. 

With that explanation, do you think we still need to change the naming here? Maybe what I described above can be better clarified in the `GZIP` docstring?